src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp

```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
                     case EVENT_ENRAGE:
-                        /// @todo Add missing text
+                        if (state == STATE_GLUTH_EATING)
+                        {
+                            events.ScheduleEvent(EVENT_ENRAGE, 4 * IN_MILLISECONDS);
+                            break;
+                        }
+
+                        Talk(EMOTE_ENRAGE);
                         DoCast(me, SPELL_ENRAGE);
-                        events.ScheduleEvent(EVENT_ENRAGE, 15000);
+                        events.ScheduleEvent(EVENT_ENRAGE, urand(16, 22) *IN_MILLISECONDS);
                         break;
                     case EVENT_DECIMATE:
-                        /// @todo Add missing text
-                        DoCastAOE(SPELL_DECIMATE);
-                        events.ScheduleEvent(EVENT_DECIMATE, 105000);
+                        if (state == STATE_GLUTH_EATING)
+                        {
+                            events.ScheduleEvent(EVENT_DECIMATE, 5 * IN_MILLISECONDS);
+                            break;
+                        }
+
+                        Talk(EMOTE_DECIMATE);
+                        DoCast(SPELL_DECIMATE);
+                        for (int i = 1; i <= 15; i++)
+                            events.ScheduleEvent(EVENT_SEARCH_ZOMBIE_MULTI, 3*i * IN_MILLISECONDS);
+                        events.ScheduleEvent(EVENT_DECIMATE, urand(110, 135) * IN_MILLISECONDS);
                         break;
                     case EVENT_BERSERK:
+                        Talk(EMOTE_BERSERKER);
                         DoCast(me, SPELL_BERSERK);
-                        events.ScheduleEvent(EVENT_BERSERK, 5*60000);
+                        events.ScheduleEvent(EVENT_BERSERK, 5 * MINUTE*IN_MILLISECONDS); //refresh the hard enrage buff
                         break;
                     case EVENT_SUMMON:
-                        for (int32 i = 0; i < RAID_MODE(1, 2); ++i)
-                            DoSummon(NPC_ZOMBIE, PosSummon[rand32() % RAID_MODE(1, 3)]);
-                        events.ScheduleEvent(EVENT_SUMMON, 10000);
+                        if (Is25ManRaid()) // one wave each 10s. one wave=1 zombie in 10man and 2 zombies in 25man.
+                        {
+                            DoSummon(NPC_ZOMBIE, PosSummon[rand32() % 3]);
+                            DoSummon(NPC_ZOMBIE, PosSummon[rand32() % 3]);
+                        }
+                        else
+                            DoSummon(NPC_ZOMBIE, PosSummon[0]);
+                        events.ScheduleEvent(EVENT_SUMMON, 10 * IN_MILLISECONDS);
+                        break;
+                    case EVENT_SEARCH_ZOMBIE_SINGLE:
+                    {
+                        Creature* zombie = nullptr;
+                        for (SummonList::const_iterator itr = summons.begin(); !zombie && itr != summons.end(); ++itr)
+                            if (ObjectAccessor::GetCreature(*me, *itr)->IsWithinDistInMap(me, 10.0))
```

Comment:
Cache ObjectAccessor::GetCreature returns, please.
Also, there is no core guarantee that guids are still valid at any given time, so nullptr check this.